<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns:MadCap="http://www.madcapsoftware.com/Schemas/MadCap.xsd" MadCap:ignoredWords="as to;INIA;website;MIAMExpress;arrays;microarrays;Affymetrix;CodeLink;dChip;RMA;PDNN;VSN;GCRMA;Quantiles;Qspline;quantile;loessScale;loessQuantile;cf;limma;vsn;datasets" MadCap:status="Complete" MadCap:lastBlockDepth="4" MadCap:lastHeight="1305" MadCap:lastWidth="1203" MadCap:disableMasterStylesheet="true" MadCap:tocPath="Analyzing Gene Lists" MadCap:InPreviewMode="false" MadCap:RuntimeFileType="Topic" MadCap:TargetType="WebHelp" MadCap:PathToHelpSystem="../" MadCap:HelpSystemFileName="PhenoGen_Overview.xml" MadCap:SearchType="Stem">
    <head>
        <link href="SkinSupport/MadCap.css" rel="stylesheet" type="text/css" /><title>Annotation Overview</title>
        <link href="Resources/Stylesheets/PhenoGen.css" rel="stylesheet" type="text/css" />
        <script src="SkinSupport/MadCapAll.js" type="text/javascript">
        </script>
    </head>
    <body>
        <p class="MCWebHelpFramesetLink" style="display: none;"><a href="../PhenoGen_Overview_CSH.htm#Annotation_Overview.htm" style="">Open topic with navigation</a>
        </p>
        <h1 class="pgBrkBefore"><a name="kanchor85"></a>Annotating Gene Lists</h1>
        <p>Gene lists on the PhenoGen website can contain identifiers from any 
 source and can be translated using a tool called iDecoder. This tool translates 
 identifiers to and from the following identifier types:</p>
        <ul>
            <li class="Bullet" value="1">AFFYMETRIX 3' PROBESET 
 ID</li>
            <li class="Bullet" value="2">AFFYMETRIX EXON PROBESET ID</li>
            <li class="Bullet" value="3">AFFYMETRIX EXON TRANSCRIPT ID</li>
            <li class="Bullet" value="4">CODELINK PROBE ID</li>
            <li class="Bullet" value="5">ENSEMBL ID</li>
            <li class="Bullet" value="6">ENTREZ GENE ID</li>
            <li class="Bullet" value="7">FLYBASE ID</li>
            <li class="Bullet" value="8">GENE SYMBOL</li>
            <li class="Bullet" value="9">MGI ID</li>
            <li class="Bullet" value="10">REFSEQ PROTEIN ID</li>
            <li class="Bullet" value="11">REFSEQ GENE ID</li>
            <li class="Bullet" value="12">RGD ID</li>
            <li class="Bullet" value="13">SWISSPROT 
 ID</li>
            <li class="Bullet" value="14">UNIGENE ID</li>
        </ul>
        <p>The data used to translate these values comes from information downloaded 
 from the following organizations:</p>
        <ul>
            <li class="Bullet" value="1">Affymetrix</li>
            <li class="Bullet" value="2">CodeLink (GE Healthcare)</li>
            <li class="Bullet" value="3">Ensembl</li>
            <li class="Bullet" value="4">FlyBase</li>
            <li class="Bullet" value="5">MGI</li>
            <li class="Bullet" value="6">NCBI</li>
            <li class="Bullet" value="7">RGD</li>
            <li class="Bullet" value="8">Swissprot</li>
        </ul>
        <p>Your uploaded gene lists can be a mixture of many ID types, but all 
 identifiers in a gene list must be for the same organism. For example, 
 you can upload a gene list that contains an Affymetrix probe set ID, an 
 official gene symbol, an Entrez Gene ID, and a RefSeq Gene ID, and iDecoder 
 translates them into the identifier types appropriate for the selected 
 tool.</p>
        <p>iDecoder is the underlying program for the annotation tools on the PhenoGen 
 website that maps gene identifiers between databases and . For instance, 
 if database 1 contains entry A and database 2 contains entry B, and both 
 A and B refer to entry C 
 in database 3, but not to each other, iDecoder identifies that A and B 
 are related. The method is very efficient in unearthing previously unknown 
 equivalent IDs. </p>
        <p>There are two levels of annotation available on the PhenoGen website:</p>
        <ul>
            <li class="Bullet" value="1"><a href="#Basic" target="" title="" alt="" class="MCXref" xrefformat="{quote}{paratext}{quote}">"Basic Annotation"</a>
            </li>
            <li class="Bullet" value="2"><a href="#Advanced" target="" title="" alt="" class="MCXref" xrefformat="{quote}{paratext}{quote}">"More Annotation"</a>
            </li>
        </ul>
        <h2 style="background-color: Transparent;"><a name="Basic"></a><a name="kanchor86"></a>Basic Annotation</h2>
        <p>Basic annotation displays links to the most popular databases for each 
 of the identifiers in a gene list. In addition to general annotation, 
 the Basic Annotation tool also provides information on expression QTL 
 (eQTL), based on mouse or rat data. A <span style="font-weight: bold;">QTL</span> 
 column displays in the Basic Annotation table and a <span style="font-weight: bold;">PhenoGen 
 eQTL</span> link displays when the gene list entry matches either a probe 
 ID or gene symbol in the eQTL data. </p>
        <h3><a name="kanchor87"></a><a name="kanchor88"></a><a name="kanchor89"></a><a name="kanchor90"></a><a name="eQTL"></a>Expression QTL (eQTL)</h3>
        <p>The purpose of expression QTLs is to determine the location in the genome 
 that controls the transcription level of a gene. eQTLs are calculated 
 using traditional QTL techniques where the quantitative trait of interest 
 is the expression level of a gene as measured by microarray analysis. 
 On the PhenoGen website, eQTLs have been calculated for both mouse and 
 rat data. When you run Basic Annotation on gene lists for either of these 
 species, the eQTLs are reported. <a href="Expression_QTL_Derivation.htm" target="" title="" alt="" class="MCXref" xrefformat="See {quote}{paratext}{quote}">See "Expression QTL Derivation"</a>.</p>
        <p>In the expression QTL table, the physical location of the probe set 
 ID is shown, along with the location of the marker that represents the 
 maximum LOD score for that transcript. The location of the marker with 
 the maximum LOD score indicates the region of transcriptional control. 
 For this analysis, if the physical location of a gene is near the location 
 of transcriptional control, the gene is considered to be cis (locally)-regulated. 
 Otherwise, the gene is considered trans (distally)-regulated. The physical 
 location of probe(set)s were obtained using the BLAT software (<a href="http://www.kentinformatics.com" target="_blank">http://www.kentinformatics.com</a>) and mapping on the 
 NCBI m37 mouse genome assembly and the RGSCv3.4 genome assembly obtained 
 from the UCSC Genome Browser (<a href="http://genome.ucsc.edu" target="_blank">http://genome.ucsc.edu</a>).</p>
        <h3><a name="Allen_Brain_Atlas"></a>Allen Brain Atlas</h3>
        <p>You can obtain the regional expression pattern of a given gene in the 
 brain of a C57BL/6J mouse by clicking on the link provided in the <span style="font-weight: bold;">Allen Brain Atlas</span> column in the Basic 
 Annotation table. These links are available ONLY for mouse gene lists. 
 The Allen Brain Atlas (Lein et al., 2007) is an open-access database of 
 gene expression in the C57BL/6J brain tissue. This database was created 
 by the Allen Institute for Brain Science (Seattle, Washington) and contains 
 data for genome-wide RNA expression obtained using high-throughput, <span style="font-style: italic;">in situ</span> hybridization. In addition 
 to the expression data, the Atlas also has a number of tools available 
 for analyzing and visualizing the <span style="font-style: italic;">in 
 situ</span> images. Click the <span style="font-weight: bold;">Instructions</span> 
 link, at the top of the Allen Brain Atlas column, to see basic instructions 
 for viewing images on the Allen Brain Atlas. Comprehensive help documentation 
 is available online at: <a href="http://www.brain-map.org">http://www.brain-map.org</a>. 
 </p>
        <h5>References</h5>
        <ol>
            <li class="Numbered" value="1">Allen Brain Atlas [Internet]. 
 Seattle (WA): Allen Institute for Brain Science. © 2004-2007. Available 
 from: http://www.brain-map.org .</li>
            <li class="Numbered" value="2">Lein ES et al. (2007). 
 Genome-wide atlas of gene expression in the adult mouse brain. Nature 
 445:168-176, doi:10.1038/nature05453. </li>
            <li class="Numbered" value="3">Markram H (2007). Industrializing 
 neuroscience. Nature 445:160-161.</li>
        </ol>
        <h2><a name="Advanced"></a><a name="kanchor91"></a>More Annotation</h2>
        <p>More annotation options allow you to select a customized set of databases 
 for the annotation of a given gene list. After you select a gene list, 
 you can perform annotation and select one or more of the different databases 
 to obtain further annotation. You can download this information. <a href="Using_More_Annotation_Options.htm" target="" title="" alt="" class="MCXref" xrefformat="See {quote}{paratext}{quote}">See "Using More Annotation Options"</a> for details.</p>
        <p class="SeeAlso" MadCap:conditions="Primary.Online">See Also</p>
        <p MadCap:conditions="Primary.Online"><a href="Performing_Annotation.htm" MadCap:conditions="Primary.Online">Performing 
 Annotation</a>
        </p>
        <p MadCap:conditions="Primary.Online"><a href="Using_More_Annotation_Options.htm">Using More Annotation Options</a>
        </p>
        <script type="text/javascript" src="SkinSupport/MadCapBodyEnd.js">
        </script>
    </body>
</html>