<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns:MadCap="http://www.madcapsoftware.com/Schemas/MadCap.xsd" MadCap:lastBlockDepth="4" MadCap:lastHeight="1401" MadCap:lastWidth="1222" MadCap:disableMasterStylesheet="true" MadCap:tocPath="Analyzing Microarrays|Analyzing Datasets|Filtering" MadCap:InPreviewMode="false" MadCap:RuntimeFileType="Topic" MadCap:TargetType="WebHelp" MadCap:PathToHelpSystem="../" MadCap:HelpSystemFileName="PhenoGen_Overview.xml" MadCap:SearchType="Stem">
    <head><title>CodeLink Gene Filtering Procedures</title>
        <link href="Resources/Stylesheets/PhenoGen.css" rel="stylesheet" type="text/css" />
        <script src="SkinSupport/MadCapAll.js" type="text/javascript">
        </script>
    </head>
    <body>
        <p class="MCWebHelpFramesetLink" style="display: none;"><a href="../PhenoGen_Overview_CSH.htm#CodeLink_Gene_Filtering_Procedures.htm" style="">Open topic with navigation</a>
        </p>
        <h1><a name="kanchor61"></a>CodeLink Gene Filtering Procedures</h1>
        <h2><a name="Control_Gene_Filter_Codelink"></a>CodeLink Control Gene Filter</h2>
        <p>There are three basic types of CodeLink probes that are used in the 
 Control Gene Filter:</p>
        <ul>
            <li class="Bullet" value="1"><span style="font-weight: bold;">(D) 
 Discovery</span> - The probes corresponding to the genes of interest for 
 a particular species.</li>
            <li class="Bullet" value="2"><span style="font-weight: bold;">(P) 
 Positive controls</span> - The bacterial probes corresponding to bacterial 
 transcripts that are spiked in at the total RNA level and are used to 
 evaluate the sensitivity and dynamic range of the platform.</li>
            <li class="Bullet" value="3"><span style="font-weight: bold;">(N) 
 Negative controls</span> - The bacterial probes used for evaluating the 
 degree of non-specific assay background and negative control threshold.</li>
        </ul>
        <p>When this filter is complete, both P and N 
 are removed.</p>
        <h2><a name="CodeLink Call Filter"></a>CodeLink Call Filter</h2>
        <p>Use the <span style="font-style: italic;">CodeLink Call Filter</span> 
 to remove internal control probes from the analysis. The CodeLink Call 
 filter is based on the quality control flag results measured from the 
 imaging software; <span style="font-style: italic;">CodeLink Expression 
 v4.1 algorithm</span>. The flags are:<br />&#160;</p>
        <table style="border-spacing: 0px;border-spacing: 0px;" cellspacing="0">
            <col />
            <col />
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;background-color: #eeeeee;border-left-style: none;border-top-style: Solid;border-top-color: #c0c0c0;border-right-style: Solid;border-right-color: #c0c0c0;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-right-width: 1px;border-top-width: 1px;border-bottom-width: 1px;" bgcolor="#EEEEEE">
                    <p class="TableHeader">Flag</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;background-color: #eeeeee;border-top-style: Solid;border-top-color: #c0c0c0;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-top-width: 1px;border-bottom-width: 1px;" bgcolor="#EEEEEE">
                    <p class="TableHeader">Description</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #c0c0c0;border-right-width: 1px;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;">
                    <p class="TableText">G</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;">
                    <p class="TableText">The spot has passed all quality control measures and 
 is defined as good.</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;border-right-color: #c0c0c0;border-right-width: 1px;border-left-style: none;">
                    <p class="TableText">M</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;border-right-style: none;">
                    <p class="TableText">The spot is identified in the MSR (Manufacturing Spot 
 Removed) File and no intensity data is provided. The probe was masked 
 after printing because it represented a suboptimal probe. Data from these 
 spots is disregarded.</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;border-right-color: #c0c0c0;border-right-width: 1px;border-left-style: none;">
                    <p class="TableText">C</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;border-right-style: none;">
                    <p class="TableText">The spot has a high level of background contamination. 
 Its background is above the global background population.</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;border-right-color: #c0c0c0;border-right-width: 1px;border-left-style: none;">
                    <p class="TableText">I</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;border-right-style: none;">
                    <p class="TableText">The spot has an irregular shape.</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;border-right-color: #c0c0c0;border-right-width: 1px;border-left-style: none;">
                    <p class="TableText">L</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;border-right-style: none;">
                    <p class="TableText">The spot has a near background signal.</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-bottom-style: Solid;border-right-color: #c0c0c0;border-right-width: 1px;border-left-style: none;border-bottom-color: #c0c0c0;border-bottom-width: 1px;">
                    <p class="TableText">S</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-bottom-style: Solid;border-bottom-color: #c0c0c0;border-bottom-width: 1px;border-right-style: none;">
                    <p class="TableText">The spot has a high number of saturated pixels, typically 
 above 60,000 units.</p>
                </td>
            </tr>
        </table>
        <p>A spot may receive more than one quality control flag. For example, 
 it may be labeled CI if it has both background contamination and an irregular 
 shape.</p>
        <p>For the filter, you can specify the number of samples with either a 
 G (Good) flag or an L (Low) flag needed, to keep or remove the probe from 
 further analysis. Samples that have any of the other flags are not counted 
 towards the number of samples needed for retention or elimination.</p>
        <h2><a name="GeneSpring Call Filter"></a>GeneSpring 
 Call Filter</h2>
        <p>The <span style="font-style: italic;">GeneSpring Call Filter</span> 
 uses present/absent calls generated from the imaging software, which was 
 designed to mimic the present/absent calls generated by GeneSpring software 
 for Affymetrix data. The P (present), A (absent), M (marginal), and U 
 (unknown) calls are based on the quality control flags listed in the CodeLink 
 Call Filter.</p>
        <ul>
            <li class="Bullet" value="1"><span style="font-weight: bold;">(P) 
 Present Call</span> - The spot receives a P call if it has a G quality 
 control flag.</li>
            <li class="Bullet" value="2"><span style="font-weight: bold;">(A) 
 Absent Call</span> - The spot receives an A call if it has an L quality 
 control flag.</li>
            <li class="Bullet" value="3">(<span style="font-weight: bold;">M) 
 Marginal Call</span> - The spot receives an M call if it has a C, I, or 
 S quality control flag.</li>
            <li class="Bullet" value="4"><span style="font-weight: bold;">(U) 
 Unknown Call</span> - The spot receives a U call if it has an M quality 
 control flag.</li>
        </ul>
        <p>Like the CodeLink Call Filter, you can specify the number of samples 
 with either a P (Good) call or an A (Low) call needed to keep or remove 
 the probe from further analysis. Samples that have an M or U call are 
 not counted toward the number of samples needed for retention or elimination.</p>
        <h2><a name="Median_Filter"></a>Median Variance Filter</h2>
        <p>Use the <span style="font-style: italic;">Median Variance Filter</span> 
 to retain probes with greater variation than the median variation gene. 
 Genes whose expression does not vary are unlikely to be differentially 
 expressed. Since it is likely that most genes are not differentially expressed 
 or associated with a particular phenotype, the median variance across 
 all genes is a reasonable model of null variation; i.e., the variation 
 due to other factors.&#160;The variance (s<span style="vertical-align: Super;">2</span>) 
 is calculated across all subjects for each gene. The null hypothesis is 
 that these variances represent random and normally distributed noise.&#160;For 
 each gene, &#160;the 
 statistic <span style="font-style: italic;">W 
 = (N-1)s</span><span style="font-style: italic;vertical-align: Super;">2</span><span style="font-style: italic;">/median(s</span><span style="font-style: italic;vertical-align: Super;">2</span><span style="font-style: italic;">) </span>is computed, where <span style="font-style: italic;">N</span> 
 is the number of observations of the gene. It is approximately chi-square 
 distributed with <span style="font-style: italic;">N-1</span> degree of 
 freedom [(Rosner, 2000), p246].&#160;The system calculates a p-value for 
 rejecting the null hypothesis and performs the False Discovery Rate (FDR) 
 multiple testing correction (Benjamini 
 &amp; Hochberg 1995), setting the FDR to 10%.&#160;The result is a list 
 of genes with significantly greater variation than the median variation 
 across genes with, at most, 10% of that list including genes that have 
 true variation less than, or equal to, median variation.</p>
        <h5>References</h5>
        <ol style="list-style: decimal;list-style-type: decimal;">
            <li value="1">Fundamentals of Biostatistics. &#160;Bernard 
 Rosner, 5th edition Duxbury &#160;Thomson 
 Learning. &#160;Pacific 
 Grove, CA USA</li>
            <li value="2">Benjamini Y, Hochberg Y. 1995. Controlling the 
 false discovery rate: A practical and powerful approach to multiple testing. 
 J R Stat Soc B 57:289–300</li>
        </ol>
        <h2><a name="Coefficient FIlter"></a>Coefficient Variation Filter</h2>
        <p>Use the<span style="font-style: italic;"> Coefficient Variation Filter</span> 
 to remove probes whose variation across samples is higher than a threshold. 
 The Coefficient Variation filter measures the consistency of the probes 
 across all samples. The coefficient of variation (CV) of each probe is 
 calculated as standard deviation divided by mean. A high CV value reflects 
 inconsistency among the samples within the group. For a two-group comparison 
 study, the CV of each group is calculated independently. You can predefine 
 a cutoff value, where genes with CVs above the cutoff value are removed. 
 You can also define different cutoff thresholds to control the consistency 
 levels of the genes.</p>
        <h2><a name="Negative_Control_Probe_Filter"></a>Negative Control Filter</h2>
        <p>Use the<span style="font-style: italic;"> Negative Control Filter</span> 
 to keep or remove probes based on whether the negative control probes 
 are above or below detection limits. The negative control probe filter 
 takes advantage of the negative controls on the CodeLink arrays to create 
 thresholds for detection. Detection limits are calculated for each array 
 individually based on the negative controls on that array only. The detection 
 limit is equal to the mean plus two standard deviations of the set of 
 negative probes with the highest mean. Both the mean and standard deviation 
 are calculated based on a 5% trimmed data set. &#160;</p>
        <h2><a name="Heritability_Filter"></a>Heritability Filter</h2>
        <p style="background-color: Transparent;">Use the <span style="font-style: italic;">Heritability 
 Filter</span> &#160;to 
 limit the probes analyzed to those with a high genetic heritability. This 
 filter eliminates probes for transcripts whose environmental influence 
 on expression is high compared to the strict genetic influence. Broad 
 sense heritability has been calculated on the public HXB/BXH recombinant 
 inbred rat panel normalized using cyclic LOESS. The broad sense heritability 
 is calculated for each probe separately using an ANOVA model. Because 
 the public data sets are based on the CodeLink Whole Genome Rat Array, 
 this filter is only available for data sets on that chip. A minimum heritability 
 threshold for inclusion must be specified. All filtering is done on the 
 probe level.</p>
        <h2>eQTL/bQTL Filter</h2>
        <p>Another way to prioritize genes for analysis is to limit those that are considered to genes whose transcription levels are controlled from the same genetic region that controls the phenotype/behavior of interest (e.g., Tabakoff et al. 2009).  We have identified expression quantitative trait loci (eQTL) for probes from brain tissue of the public HXB/BXH recombinant inbred rat panel normalized using cyclic LOESS.  The user chooses a significance threshold for eQTL and the appropriate bQTL to compare to.  Probes are retained if their eQTL is significant and the location of the marker (SNP) with the maximum association with transcript expression is within the chosen bQTL limits. All filtering is done on the probe level.</p>
        <h2>Gene List Filter</h2>
        <p>Filtering 
 by Gene List allows you to select a gene list that has been previously 
 created and either keep or remove all the genes within that gene list.</p>
        <script type="text/javascript" src="SkinSupport/MadCapBodyEnd.js">
        </script>
    </body>
</html>