<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns:MadCap="http://www.madcapsoftware.com/Schemas/MadCap.xsd" MadCap:lastBlockDepth="2" MadCap:lastHeight="400" MadCap:lastWidth="1229" MadCap:disableMasterStylesheet="true" MadCap:tocPath="Analyzing Microarrays|Analyzing Datasets|Types of Statistical Analysis" MadCap:InPreviewMode="false" MadCap:RuntimeFileType="Topic" MadCap:TargetType="WebHelp" MadCap:PathToHelpSystem="../" MadCap:HelpSystemFileName="PhenoGen_Overview.xml" MadCap:SearchType="Stem">
    <head>
        <link href="SkinSupport/MadCap.css" rel="stylesheet" type="text/css" /><title>Correlation Analysis</title>
        <link href="Resources/Stylesheets/PhenoGen.css" rel="stylesheet" type="text/css" />
        <script src="SkinSupport/MadCapAll.js" type="text/javascript">
        </script>
    </head>
    <body>
        <p class="MCWebHelpFramesetLink" style="display: none;"><a href="../PhenoGen_Overview_CSH.htm#Correlation_Analysis.htm" style="">Open topic with navigation</a>
        </p>
        <h1 style="background-color: Transparent;"><a name="kanchor65"></a>Correlation Analysis</h1>
        <p>A correlation analysis is used to determine if two variables are associated 
 with each other. The Correlation Analysis tool on the PhenoGen website 
 searches for genes whose expression value correlates with a phenotype 
 value. You can use the correlation analysis tools with one of your own 
 experiments or with one of the public datasets on the website (see <a href="Viewing_Datasets.htm#publicDatasets" target="" title="" alt="" class="MCXref" xrefformat="{quote}{paratext}{quote}">"Public Datasets"</a> for details). You upload phenotype 
 data which can be any continuous measurement. </p>
        <p>For example, if you are interested in genes expressed in the brain that 
 are correlated with alcohol preference in mice, you can upload a file 
 containing alcohol preference values for any or all of the inbred or recombinant 
 inbred strains for which whole brain gene expression data is available 
 on the PhenoGen website see <a href="Viewing_Datasets.htm#publicDatasets" target="" title="" alt="" class="MCXref" xrefformat="{quote}{paratext}{quote}">"Public Datasets"</a> for details). 
 A positive correlation between alcohol preference and the expression value 
 of a transcript in strains where the expression for that transcript is 
 high might indicate a relationship between that gene and alcohol preference. 
 Similarly, a negative correlation might indicate an inverse relationship.</p>
        <h2><a name="Gene_Expresssion_and_Correlation_Analysis"></a>Statistics for 
 Correlation Analysis</h2>
        <p>For correlation analysis, mean expression values are calculated within 
 strain. These mean values are correlated with the strain phenotypic measures 
 that you upload. You can choose either a Pearson Correlation Coefficient 
 or a Spearman Rank Correlation Coefficient to calculate correlation. </p>
        <h3>Pearson Test</h3>
        <p>The Pearson correlation coefficient is a parametric test for a linear 
 relationship between two variables. A parametric test is more powerful 
 when the two variables involved are truly normally distributed, but a 
 non-parametric test is more accurate when this normality assumption is 
 not met.</p>
        <h3>Spearman Test</h3>
        <p>The Spearman rank correlation is a non-parametric test that looks at 
 the correlation of the ranks of the values rather than the actual values. 
 A non-parametric test is necessary when the distribution of either of 
 the variables is not normal. </p>
        <script type="text/javascript" src="SkinSupport/MadCapBodyEnd.js">
        </script>
    </body>
</html>