<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns:MadCap="http://www.madcapsoftware.com/Schemas/MadCap.xsd" MadCap:lastBlockDepth="6" MadCap:lastHeight="656" MadCap:lastWidth="956" MadCap:conditionTagExpression="" MadCap:disableMasterStylesheet="true" MadCap:tocPath="" MadCap:InPreviewMode="false" MadCap:RuntimeFileType="Topic" MadCap:TargetType="WebHelp" MadCap:PathToHelpSystem="../" MadCap:HelpSystemFileName="PhenoGen_Overview.xml" MadCap:SearchType="Stem">
    <head><title>Detailed Transcription Information</title>
        <link href="Resources/Stylesheets/PhenoGen.css" rel="stylesheet" type="text/css" />
        <script src="SkinSupport/MadCapAll.js" type="text/javascript">
        </script>
    </head>
    <body>
        <p class="MCWebHelpFramesetLink" style="display: none;"><a href="../PhenoGen_Overview_CSH.htm#Detailed_Transcription_Info.htm" style="">Open topic with navigation</a>
        </p>
        <h1><a name="kanchor15"></a>Detailed Transcription Information</h1>
        <p><span style="font-weight: bold;">Detailed Transcription Information</span> is available regardless of whether you are logged in. A video provides a demonstration of the functionality. </p>
        <p>On the Detailed Transcription Information page:</p>
        <ol>
            <li value="1">Enter a <b>gene identifier</b> or <b>region</b> in the gene field. For example:</li>
        </ol>
        <blockquote>
            <ul>
                <li value="1">Gene identifiers: gene symbol, probeset ID, ensembl ID, etc.</li>
                <li value="2">Regions: chr1:1-50000 (Chromosome 1 bp 1-50,000), chr1:5000+-2000 (e.g., Chromosome 1 bp 3,000-7,000), chr1:5000+2000 (e.g., Chromosome 1 bp 5,000-7,000), etc.</li>
            </ul>
        </blockquote>
        <ol>
            <p>OR</p>
            <li value="2">Click <b>Translate Region to Mouse/Rat</b> to find regions on the Mouse/Rat genome that correspond to a region of interest in the Human/Mouse/Rat genome.</li>
            <li value="3">Choose a species.</li>
            <li value="4">Click <b>Get Transcription Details</b>. If they are available, transcription details for the gene or region you entered display in a combination of interactive images and graphs.</li>
            <p>
                <img src="Images/bulb.gif" /><b>Note:</b> For rats only, whole brain RNA-Seq transcriptome reconstruction data also displays. </p>
        </ol>
        <h3><a name="Transcri"></a>Transcriptome Reconstruction (Rat Only)</h3>
        <p>The rat brain transcriptome was reconstructed using RNA-Seq data from the BN-Lx/CubPrin and SHR/OlaPrin strains.  PolyA+ RNA was extracted from three rats per strain and sequenced by an Illumina HiSeq2000 using paired-end reads 100 base pairs long.  Reads were aligned to the rat genome (rn4) using TopHat (Trapnell et al 2009) after trimming each read fragment to 70 base pairs.  Almost 300 million reads were sequenced and successfully aligned to the genome.  The transcriptome reconstruction was done in each strain separately using the CuffLinks program (Trapnell et al 2010) and then merged.   Transcripts are labeled CUFF.X.Y, where X is a numeric identifier for the gene and Y is the numeric identifier for a specific isoform/transcript of that gene.  The aligned RNA-Seq data is available for download from PhenoGen as SAM files.</p>
        <p>There currently is no data available from a mouse brain transcriptome reconstruction, but the analyses are underway.</p>
        <p style="font-weight: bold;">References</p>
        <ol>
            <li value="1">Trapnell, C., Pachter, L., &amp; Salzberg, S. L. (2009). TopHat: discovering splice junctions with RNA-Seq. Bioinformatics (Oxford, England), 25(9), 1105–1111. doi:10.1093/bioinformatics/btp120</li>
            <li value="2">Trapnell, C., Williams, B. A., Pertea, G., Mortazavi, A., Kwan, G., van Baren, M. J., Salzberg, S. L., et al. (2010). Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation. Nature biotechnology, 28(5), 511–515. doi:10.1038/nbt.1621</li>
        </ol>
        <p class="SeeAlso" MadCap:conditions="Primary.Online">See Also</p>
        <p MadCap:conditions="Primary.Online"><a href="Detailed_Transcription_Info_Genes.htm">Detailed Transcription Information for Gene Identifiers</a>
        </p>
        <p MadCap:conditions="Primary.Online"><a href="Detailed_Transcription_Info_Regions.htm">Detailed Transcription Information for Regions</a>
        </p>
        <script type="text/javascript" src="SkinSupport/MadCapBodyEnd.js">
        </script>
    </body>
</html>