<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns:MadCap="http://www.madcapsoftware.com/Schemas/MadCap.xsd" MadCap:lastBlockDepth="4" MadCap:lastHeight="1311" MadCap:lastWidth="1203" MadCap:disableMasterStylesheet="true" MadCap:tocPath="Getting Started" MadCap:InPreviewMode="false" MadCap:RuntimeFileType="Topic" MadCap:TargetType="WebHelp" MadCap:PathToHelpSystem="../" MadCap:HelpSystemFileName="PhenoGen_Overview.xml" MadCap:SearchType="Stem">
    <head>
        <link href="SkinSupport/MadCap.css" rel="stylesheet" type="text/css" /><title>PhenoGen Website Overview</title>
        <link href="Resources/Stylesheets/PhenoGen.css" rel="stylesheet" type="text/css" />
        <script src="SkinSupport/MadCapAll.js" type="text/javascript">
        </script>
    </head>
    <body>
        <p class="MCWebHelpFramesetLink" style="display: none;"><a href="../PhenoGen_Overview_CSH.htm#Phenogen_Overview.htm" style="">Open topic with navigation</a>
        </p>
        <h1 style="background-color: Transparent;"><a name="kanchor4"></a>PhenoGen Website Overview</h1>
        <p>The PhenoGen website shares experimental data with a worldwide community 
 of investigators and provides a flexible, integrated, multi-resolution 
 repository of neuroscience transcriptomic genetic data for collaborative 
 research on genomic disorders. </p>
        <p>The website provides a comprehensive system to organize, query, analyze, 
 and retrieve high-throughput gene expression data, as well as providing 
 users with computational tools for integrated analysis of neuroscience 
 data, biomedical literature, gene functional annotations, and Quantitative 
 Trait Loci (QTLs). </p>
        <p>The PhenoGen website allows data to be classified as "Semi-public" 
 or "Open Access". All of the information about the data uploaded 
 at the PhenoGen website is visible to every registered user (see <a href="Registration.htm" target="" title="" alt="" class="MCXref_custom" xrefformat="{quote}{paratext}{quote}">"Registering an Account"</a> for details). Registered users have full access to data that is classified 
 as "Open Access" and do not need to obtain permission from the 
 curator (Principal Investigator) of the data. "Semi-public" 
 data can only be accessed and downloaded after the curator of the data 
 grants a user permission to do so. Registered users can use the data for 
 "in-silico" analysis or can download the data for analysis with 
 their own statistical software.</p>
        <p style="background-color: Transparent;">The website also has nine pre-compiled 
 “Public” microarray datasets that can be used and downloaded by all registered 
 users for gene expression analysis, including correlating with user-provided 
 phenotype data. These datasets include inbred and recombinant inbred mice 
 and rat strains. </p>
        <p>The PhenoGen website allows you to:</p>
        <ul>
            <li class="Bullet" value="1">Upload microarray raw data 
 into a MIAME-compliant database.</li>
            <li class="Bullet" value="2">Upload gene lists.</li>
            <li class="Bullet" value="3">Share data with other investigators 
 around the world.</li>
            <li class="Bullet" value="4">Search literature and save 
 results.</li>
            <li class="Bullet" value="5">Translate gene identifiers 
 to and from multiple databases.</li>
            <li class="Bullet" value="6">Determine phenotypic QTLs 
 using BXD recombinant inbred mice, HXB/BXH recombinant inbred rats, or 
 LXS recombinant inbred mice.</li>
            <li class="Bullet" value="7">Match physical location of 
 genes of interest and their eQTL to phenotypic QTLs.</li>
            <li class="Bullet" value="8">Correlate gene expression 
 with a phenotype.</li>
        </ul>
        <p>You can also perform:</p>
        <ul>
            <li class="Bullet" value="1">Microarray data quality control 
 analysis and normalization.</li>
            <li class="Bullet" value="2">Data filtering (noise filtering).</li>
            <li class="Bullet" value="3">Statistical analyses, including 
 the most common statistical tests and permutations.</li>
            <li class="Bullet" value="4">Promoter analysis (transcription 
 factors).</li>
            <li class="Bullet" value="5">Queries about genetic variations 
 (e.g. SNPs or polymorphisms) in the transcripts of interest.</li>
        </ul>
        <h2><a name="kanchor5"></a><a name="Website_Process_Flow"></a>Website Process Flow for Microarray Analyses&#160;</h2>
        <table style="caption-side: top;margin-left: 0;margin-right: auto;">
            <col style="width: 50%;" />
            <col style="width: 50%;" />
            <tbody>
                <tr>
                    <td valign="top">
                        <p><span style="font-weight: bold;">Analyze Microarray Data</span>
                        </p>
                        <p>The process flow for a microarray analysis is:</p>
                        <p style="font-weight: bold;margin-left: 40px;">Dataset Creation</p>
                        <p style="margin-left: 40px;"><span style="font-style: italic;">Upload 
 microarray data.</span> If you have microarray data from a lab experiment, 
 you can upload it into a MIAME-compliant database that is part of the 
 PhenoGen website. </p>
                        <ul>
                            <ol>
                                <li class="Numbered" value="1">Retrieve arrays.</li>
                                <li class="Numbered" value="2">Select and merge 
 arrays from the data repository.</li>
                                <li class="Numbered" value="3">Finalize dataset.</li>
                                <li class="Numbered" value="4">Run quality control 
 measures on the merged arrays.</li>
                                <li class="Numbered" value="5">Review quality control 
 results.</li>
                            </ol>
                        </ul>
                        <p style="font-weight: bold;margin-left: 40px;"><span style="font-weight: bold;">Dataset 
 Preparation</span>
                        </p>
                        <ul>
                            <ol start="6">
                                <li value="6">Group the arrays based on your hypothesis 
 (e.g., disease vs. control).</li>
                                <li value="7">Normalize the dataset.</li>
                            </ol>
                        </ul>
                        <p style="font-weight: bold;margin-left: 40px;">Dataset Analysis</p>
                        <ul style="list-style: decimal;" type="1">
                            <ol start="8" style="list-style: decimal;list-style-type: decimal;">
                                <li class="Numbered" value="8">Select a single normalized 
 version.</li>
                                <li class="Numbered" value="9">Select analysis type.</li>
                                <li class="Numbered" value="10">Filter genes.</li>
                                <li class="Numbered" value="11">Perform statistical 
 analysis.</li>
                                <li class="Numbered" value="12">Perform a multiple 
 testing adjustment</li>
                                <li class="Numbered" value="13">Save gene lists.</li>
                            </ol>
                        </ul>
                        <p style="font-weight: bold;margin-left: 0px;">Research Gene Lists</p>
                        <p style="margin-left: 0px;">Do one of the following to enter a gene list 
 in PhenoGen:</p>
                        <ul style="list-style: decimal;" type="1">
                            <li class="Bullet" value="1">Generate a gene list from 
 microarray analysis (see the preceding Analyze Microarray Data section.)</li>
                        </ul>
                        <p style="margin-left: 40px;">OR </p>
                        <ul style="list-style: decimal;" type="1">
                            <li class="Bullet" value="1">Upload a gene list if you 
 have an existing gene list to interpret.</li>
                        </ul>
                        <p style="margin-left: 0px;">When your gene list is on the website, use 
 the annotation, QTL, literature search, and promoter analysis tools to 
 help interpret your list of candidate genes.</p>
                    </td>
                    <td>
                        <img src="Images/Process_Flow1_346x692.gif" style="width: 346; height: 692;" />
                    </td>
                </tr>
            </tbody>
        </table>
        <script type="text/javascript" src="SkinSupport/MadCapBodyEnd.js">
        </script>
    </body>
</html>