<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns:MadCap="http://www.madcapsoftware.com/Schemas/MadCap.xsd" MadCap:ignoredWords="as to;INIA;website;MIAMExpress;arrays;microarrays;Affymetrix;CodeLink;dChip;RMA;PDNN;VSN;GCRMA;Quantiles;Qspline;quantile;loessScale;loessQuantile;cf;limma;vsn;datasets" MadCap:lastBlockDepth="4" MadCap:lastHeight="888" MadCap:lastWidth="845" MadCap:disableMasterStylesheet="true" MadCap:tocPath="Supplementary Information" MadCap:InPreviewMode="false" MadCap:RuntimeFileType="Topic" MadCap:TargetType="WebHelp" MadCap:PathToHelpSystem="../" MadCap:HelpSystemFileName="PhenoGen_Overview.xml" MadCap:SearchType="Stem">
    <head><title>Promoter Analysis Tools</title>
        <link href="Resources/Stylesheets/PhenoGen.css" rel="stylesheet" type="text/css" />
        <script src="SkinSupport/MadCapAll.js" type="text/javascript">
        </script>
    </head>
    <body>
        <p class="MCWebHelpFramesetLink" style="display: none;"><a href="../PhenoGen_Overview_CSH.htm#Promoter_Analysis_Tools.htm" style="">Open topic with navigation</a>
        </p>
        <h1><a name="kanchor133"></a>Promoter Analysis Tools</h1>
        <table style="border-spacing: 0px;margin-left: 0;margin-right: 0;caption-side: top;width: 100%; font-size: 8pt;" cellspacing="0">
            <col />
            <col />
            <col />
            <col />
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;background-color: #c0c0c0;border-left-style: none;border-top-style: Solid;border-top-color: #808080;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-right-width: 1px;border-top-width: 1px;border-bottom-width: 1px;" bgcolor="#C0C0C0">
                    <p class="TableHeader" style="font-size: 8pt;">Program</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;background-color: #c0c0c0;border-top-style: Solid;border-top-color: #808080;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-right-width: 1px;border-top-width: 1px;border-bottom-width: 1px;" bgcolor="#C0C0C0">
                    <p class="TableHeader" style="font-size: 8pt;">Operating Principle</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;background-color: #c0c0c0;border-top-style: Solid;border-top-color: #808080;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-right-width: 1px;border-top-width: 1px;border-bottom-width: 1px;" bgcolor="#C0C0C0" width="483px">
                    <p class="TableHeader" style="font-size: 9pt;"><span style="font-size: 8pt;">Technical Data and </span><span style="font-size: 9pt;"><span style="font-size: 8pt;">URL</span></span>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;background-color: #c0c0c0;border-top-style: Solid;border-top-color: #808080;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-top-width: 1px;border-bottom-width: 1px;" bgcolor="#C0C0C0">
                    <p class="TableHeader" style="font-size: 8pt;">Reference</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-right-width: 1px;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">AIignACE</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-right-width: 1px;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Gibbs 
 sampling algorithm that returns a series of motifs as weight matrices 
 that are over-represented in the input set.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-right-width: 1px;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">Judges alignments sampled during the course 
 of the algorithm using a maximum a priori likelihood score, which gauges 
 the degree of over-representation. Provides an adjunct measure (group 
 specificity score) that takes into account the sequence of the entire 
 genome and highlights those motifs found preferentially in association 
 with the genes under consideration.</p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://atlas.med.harvard.edu/" target="_blank" style="font-size: 8pt;">http://atIas.med.harvard.edu</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">1</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">ANN-Spec</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">ModeIs 
 the DNA binding specificity of a transcription factor using a weight matrix.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">Objective function based on log likelihood 
 that a transcription factor binds at least once in each sequence of the 
 positive training data compared with the number of times it is estimated 
 to bind in the background training data. Parameter fitting is accomplished 
 with a gradient descent method, which includes Gibbs 
 sampling of the positive training examples.</p>
                    <p class="TableText" style="font-size: 9pt;"><a href="http://www.cbs.dtu.dk/services/DNAarray/ann-spec.php" target="_blank" style="font-size: 8pt;">http://www.cbs.dtu.dk/services/DNAarray/ann-spec.php</a>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">2</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">Consensus</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Models motifs using weight matrices searching 
 for lhe matrix 
 with maximum information content.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">Uses a greedy method, first finding the pair 
 of sequences that share the motif with greatest information content, then 
 finding the third sequence that can be added to the motif, resuIting in 
 greatest information content.</p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://bifrost.wustl.edu/consensus/" target="_blank" style="font-size: 8pt;">http://bifrost.wustl.edu/consensus/</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">3</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">GLAM</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Gibbs 
 sampling-based algorithm that automatically optimizes the alignment width 
 and evaluates the statistical significance of its output.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">Since the basic algorithm cannot find multiple 
 motif instances per sequence, long sequences are fragmented into shorter 
 ones, and the alignment is transformed into a weight matrix and used to 
 scan the sequences to obtain the final site predictions.</p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://zlab.bu.edu/glam/" target="_blank" style="font-size: 8pt;">http://zlab.bu.edu/glam/</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">4</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">Improbizer</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Uses expectation maximization to determine 
 weight matrices of DNA motifs that occur improbably often in the input 
 sequences.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">As a background (null) model it uses up to 
 a second-order Markov 
 model of background sequence. Optionally, Improbizer 
 constructs a Gaussian model of motif placement so that motifs that occur 
 in similar positions in the input sequences are more likely to be found.</p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://www.soe.ucsc.edu/~kent/improbizer" target="_blank" style="font-size: 8pt;">http://www.soe.ucsc.edu/~kent/improbizer</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">5</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">MEME</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Optimizes the E-value 
 of a statistic related to the information content of the motif.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">Rather than sum of information content of each 
 motif column, the statistic used is the product of the p-values of column 
 information contents. The motif search consists of performing expectation 
 maximization from starting points derived from each subsequence occurring 
 in the input sequences. MEME 
 differs from MEME3 mainly in using a correction factor to improve the 
 accuracy of the objective function.</p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://meme.sdsc.edu/" target="_blank" style="font-size: 8pt;">http://meme.sdsc.edu</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">6</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">MITRA</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Uses an efficient data structure to traverse 
 the space of IUPAC 
 patterns.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">For each pattern, MITRA 
 computes the hypergeometric score of the occurrences in the target sequences 
 relative to the background sequence and reports the highest scoring patterns.</p>
                    <p class="TableText" style="font-size: 9pt;"><a href="http://www.ccls.columbia.edu/compbio/mitra/" target="_blank" style="font-size: 8pt;">http://www.ccls.columbia.edu/compbio/mitra/</a>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">7</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">MotifSampler</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Matrix-based motif finding algorithm that extends 
 Gibbs sampling 
 by modeling the background with a higher order Markov 
 model. </p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">The probabilistic framework is further exploited 
 to estimate the expected number of motif instances in the sequence.</p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://www.esat.kuleuven.ac.be/~dna/Biol/Software.html" target="_blank" style="font-size: 8pt;">http://www.esat.kuleuven.ac.be/~dna/Biol/Software.html</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">8</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="margin-top: 2pt;margin-bottom: 2pt;font-size: 8pt;">Oligo/dyad 
 anaIysis</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Detects over-represented oligonucleotides with 
 oligo analysis and spaced motifs with dyad analysis.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">These algorithms detect statistically significant 
 motifs by counting the number of occurrences of each word or dyad and 
 comparing these with expectation. The most crucial parameter is the choice 
 of an appropriate probabilistic model for the estimation of occurrence 
 significance. </p>
                    <p class="TableText" style="font-size: 9pt;"><a href="http://rsat.ulb.ac.be/rsat/" target="_blank" style="font-size: 8pt;">http://rsat.ulb.ac.be/rsat/</a>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">9, 10</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">QuickScore</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Based on an exhaustive searching algorithm 
 that estimates probabilities of rare or frequent words in genomic texts.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">Incorporates an extended consensus method allowing 
 well-defined mismatches and uses mathematical expressions for efficiently 
 computing z-scores and p-values depending on the statistical models used 
 in their range of applicability. Special attention is paid to the drawbacks 
 of numerical instability. The background model is Markovian, 
 with order up to 3. </p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://algo.inria.fr/dolley/QuickScore/" target="_blank" style="font-size: 8pt;">http://algo.inria.fr/dolley/QuickScore/</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">11</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">SeSiMCMC</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Modification of Gibbs 
 sampler aIgorithm that models the motif as a weight matrix, optionally 
 with the symmetry of a palindrome or of a direct repeat and optionally, 
 with spacers.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">Includes two alternating stages. The first 
 one optimizes the weight matrix for a given motif and spacer length. The 
 algorithm changes the positions of the motif occurrence in the sequences 
 and infers the motif model from the current occurrences. These changes 
 are used to optimize the likelihood of sequences as being segmented into 
 the (Bernoulli) background and the motif occurrences. The optimization 
 is organized via a Gibbs-like 
 Markov chain 
 that samples positions in sequences one-by-one until the Markov 
 chain converges. The second stage looks for best motif and spacer lengths 
 for obtained motif positions. It optimizes the common information content 
 of motif and of distributions of motif occurrence positions.</p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><span style="text-decoration: underline;"><a href="http://favorov.imb.ac.ru/SeSiMCMC/" target="_blank" style="font-size: 8pt;">http://favorov.imb.ac.ru/SeSiMCMC/</a></span>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">12</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">Weeder</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Consensus-based method that enumerates exhaustively 
 all the oligos up to maximum length and collects their occurrences (with 
 substitutions) from input sequences. </p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">Each motif is evaluated according to the number 
 of sequences in which it appears and how well conserved it is in each 
 sequence, with respect to expected values derived from the oligo frequency 
 analysis of all the available upstream sequences of the same organism. 
 Different combinations of canonical motif parameters derived from the 
 analysis of known instances of yeast transcription factor binding sites 
 (length ranging from 6 to l2, number of substitutions from 1 to 4) are 
 automatically tried by the algorithm in different runs. It also analyzes 
 and compares the top-scoring motifs of each run with a simple clustering 
 method to detect which ones could be more likely to correspond to transcription 
 factor binding sites. Best instances of each motif are selected from sequences 
 using a weight matrix built with sites found by consensus-based algorithms.</p>
                    <p class="TableText" style="font-size: 9pt;"><a href="http://159.149.109.9/modtools/" target="_blank" style="font-size: 8pt;">http://159.149.109.9/modtools/</a>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">13</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">YMF</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 9pt;"><span style="font-size: 8pt;">Uses an exhaustive search algorithm to find 
 motifs with the greatest </span><i style="font-size: 8pt;">z</i><span style="font-size: 8pt;">-scores.</span>
                    </p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 9pt;"><span style="font-size: 8pt;">A </span><i style="font-size: 8pt;">p</i><span style="font-size: 8pt;">-value for the </span><i style="font-size: 8pt;">z</i><span style="font-size: 8pt;">-score is 
 used to assess the significance of the motif. Motifs themselves are short 
 sequences over the IUPAC 
 alphabet with spacers ("N"s) 
 constrained to occur in the middle of the sequence.</span>
                    </p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://bio.cs.washington.edu/software.html#yms" target="_blank" style="font-size: 8pt;">http://bio.cs.washington.edu/software.html#yms</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">14</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">Composite Module Analyst (CMA)</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Uses a multi-component fitness function for 
 selection of the promoter model which fits best to the observed gene expression 
 profile.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">Defines a promoter model based on composition 
 of transcription factor binding sites and their pairs. Adjusts the results 
 of the fitness function using a genetic algorithm for the analysis of 
 functionally related or coexpressed 
 genes.</p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://www.gene-regulation.com/cgi-bin/CMA/cma.cgi" target="_blank" style="font-size: 8pt;">http://www.gene-regulation.com/cgi-bin/CMA/cma.cgi</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">15</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">REDUCE</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Motif-based regression method for microarray 
 analysis.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">The only required inputs are (i) a single genome-wide 
 set of absolute or relative mRNA abundances and (ii) the DNA sequence 
 of the regulatory region associated with each gene that is probed. REDUCE 
 uses unbiased statistics to identify oligonucleotide motifs whose occurrence 
 in the regulatory region of a gene correlates with the level of mRNA expression. 
 Regression analysis is used to infer the activity of the transcriptional 
 module associated with each motif.</p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><u style="text-decoration: underline;"><a href="http://bussemaker.bio.columbia.edu/reduce/" target="_blank" style="font-size: 8pt;">http://bussemaker.bio.columbia.edu/reduce/</a></u>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">16</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">MotifRegressor</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Combines the advantages of matrix-based motif 
 finding and oligomer 
 motif-expression regression analysis.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 8pt;">MotifRegressor 
 first constructs candidate motifs and then applies regression analysis 
 to select motifs that are strongly correlated with changes in gene expression. 
 It is particularly effective in discovering expression-mediating motifs 
 of medium-to-long width with multiple degenerate positions. MotifRegressor 
 relies in part on MDScan, 
 a software package developed by the Brutlag 
 Lab at Stanford University.</p>
                    <p class="TableText" style="font-size: 9pt;"><a href="http://www.math.umass.edu/~conlon/mr.html" target="_blank" style="font-size: 8pt;">http://www.math.umass.edu/~conlon/mr.html</a>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">17</p>
                </td>
            </tr>
            <tr valign="top">
                <td style="padding-right: 10px;padding-left: 10px;border-left-style: none;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">CisModule</p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="191px">
                    <p class="TableText" style="font-size: 8pt;">Employs a hierarchical mixture approach to 
 model the cis-regulatory 
 module structure.</p>
                </td>
                <td rowspan="1" colspan="1" style="padding-right: 10px;padding-left: 10px;border-right-style: Solid;border-right-color: #808080;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;border-right-width: 1px;" width="483px">
                    <p class="TableText" style="font-size: 9pt;"><span style="font-size: 8pt;">It is based on the hierarchical mixture model, 
 followed by a</span><i style="font-size: 8pt;">de 
 novo</i><span style="font-size: 8pt;"> motif-module 
 discovery algorithm using the Bayesian inference of module locations and 
 within-module motif sites. Dynamic programming-like recursions are developed 
 to reduce the computational complexity from exponential to linear in sequence 
 length.</span>
                    </p>
                    <p class="TableText" style="color: #0000ff;font-size: 9pt;"><span style="text-decoration: underline;"><a href="http://www.stat.ucla.edu/~zhou/CisModule/index.html" target="_blank" style="font-size: 8pt;">http://www.stat.ucla.edu/~zhou/CisModule/index.html</a></span>
                    </p>
                </td>
                <td style="padding-right: 10px;padding-left: 10px;border-right-style: none;border-bottom-style: Solid;border-bottom-color: #808080;border-bottom-width: 1px;">
                    <p class="TableText" style="font-size: 8pt;">18</p>
                </td>
            </tr>
        </table>
        <p style="font-size: 8pt;margin-top: 2pt;margin-bottom: 2pt;">&#160;</p>
        <p class="MsoNormal"><b style="font-weight: bold;">References</b>
        </p>
        <ol start="1" style="list-style: decimal;margin-top: 0in;list-style-type: decimal;">
            <li class="kadov-p" value="1">Hughes JD, Estep 
 PW, Tavazoie 
 S, Church GM (2000). Computational identification of <i>cis</i>-regulatory 
 elements associated with functionally coherent groups of genes in <i>Saccharomyces 
 cerevisiae</i>. 
 <i style="font-style: normal;">J 
 Mol Biol</i> 
 296:1205–1214. </li>
            <li class="kadov-p" value="2">Workman CT and Stormo 
 GD (2000). ANN-Spec: 
 a method for discovering transcription factor binding sites with improved 
 specificity. <i><span style="font-style: normal;">Pacific</span><span style="font-style: normal;">Symposium on Biocomputing</span></i> 
 (ed. Altman 
 R, Dunker AK, 
 Hunter L, Klein 
 TE). 467–478 (Stanford University, Stanford, CA). </li>
            <li class="kadov-p" value="3">Hertz GZ and Stormo 
 GD (1999). Identifying DNA and protein patterns with statistically significant 
 alignments of multiple sequences. <i style="font-style: normal;">Bioinformatics</i> 
 15:563–577. </li>
            <li class="kadov-p" value="4">Frith 
 MC, Hansen 
 U, Spouge JL, 
 Weng Z (2004). 
 Finding functional sequence elements by multiple local alignment. <i style="font-style: normal;">Nucleic 
 Acids Res</i> 32:189–200. </li>
            <li class="kadov-p" value="5">Ao 
 W, Gaudet 
 J, Kent WJ, 
 Muttumu S, 
 Mango SE (2004). Environmentally induced foregut 
 remodeling by PHA-4/FoxA 
 and DAF-12/NHR. 
 <i style="font-style: normal;">Science</i> 305:1743–1746. </li>
            <li class="kadov-p" value="6">Bailey TL and Elkan C (1995). The value of prior 
 knowledge in discovering motifs with MEME. 
 <i><span style="font-style: normal;">Proceedings of the Third International</span><span style="font-style: italic;font-weight: bold;"> <![CDATA[ ]]></span><span style="font-style: normal;">Conference</span><span style="font-weight: bold;font-style: italic;"> <![CDATA[ ]]></span><span style="font-style: normal;">on 
 Intelligent Systems</span><span style="font-weight: bold;font-style: italic;"> <![CDATA[ ]]></span><span style="font-style: normal;">for Molecular Biology.</span></i> 
 21–29 (AAAI 
 Press, Menlo 
 Park, CA). </li>
            <li class="kadov-p" value="7">Eskin 
 E and Pevzner 
 P (2001). Finding composite regulatory patterns in DNA sequences. <i style="font-style: normal;">Bioinformatics</i> 
 (Supplement 1) 18:S354–S363. </li>
            <li class="kadov-p" value="8">Thijs 
 G, <i>et al <span style="font-style: normal;">(2001)</span></i>. A higher-order 
 background model improves the detection of promoter regulatory elements 
 by Gibbs sampling. 
 <i style="font-style: normal;">Bioinformatics</i> 17:1113–1122. </li>
            <li class="kadov-p" value="9">van Helden J, 
 Andre B, Collado-Vides 
 J (1998). Extracting regulatory sites from the upstream region of yeast 
 genes by computational analysis of oligonucleotide frequencies. <i style="font-style: normal;">J 
 Mol Biol</i> 
 281:827–842. </li>
            <li class="kadov-p" value="10">van Helden 
 J, Rios AF, 
 Collado-Vides 
 J (2000). Discovering regulatory elements in noncoding 
 sequences by analysis of spaced dyads. <i style="font-style: normal;">Nucleic 
 Acids Res</i> 28:1808–1818. </li>
            <li class="kadov-p" value="11">Régnier 
 M and Denise 
 A (2004). Rare events and conditional events on random strings. <i style="font-style: normal;">Discrete 
 Math Theor 
 Comput Sci</i> 
 6:191–214. </li>
            <li class="kadov-p" value="12">Favorov 
 AV, Gelfand 
 MS, Gerasimova 
 AV, Mironov AA, Makeev 
 VJ (2004). Gibbs 
 sampler for identification of symmetrically structured, spaced DNA motifs 
 with improved estimation of the signal length and its validation on the 
 ArcA binding 
 sites. <i style="font-style: normal;">Proceedings of BGRS 
 2004</i> (BGRS, 
 Novosibirsk). 
 </li>
            <li class="kadov-p" value="13">Pavesi 
 G, Mereghetti 
 P, Mauri G, 
 and Pesole 
 G (2004). Weeder Web: discovery of transcription factor binding sites 
 in a set of sequences from co-regulated genes. <i style="font-style: normal;">Nucleic 
 Acids Res 32:</i>W199–W203. </li>
            <li class="kadov-p" value="14">Sinha S and Tompa 
 M (2003). YMF: 
 a program for discovery of novel transcription factor binding sites by 
 statistical overrepresentation. 
 <i><span style="font-style: normal;">Nucleic Acids Res 31:</span></i><span style="font-style: normal;">3586–3588.</span></li>
            <li class="kadov-p" value="15">Konovalova T, Valeev T, Cheremushkin E, Kel AE (2005). 
 Composite Module Analyst: Tool for Prediction of DNA Transcription Regulation. 
 Testing on Simulated Data. ICNC 
 2:1202-1205.</li>
            <li class="kadov-p" value="16">Roven 
 C and Bussemaker 
 HJ (2003). REDUCE: an online tool for inferring cis-regulatory 
 elements and transcriptional module activities from microarray data. Nucleic 
 Acids Research 31(13):3487-3490.</li>
            <li class="kadov-p" value="17">Conlon EM, Liu XS, Lieb JD, Liu JS(2003). Proc 
 Natl Acad 
 Sci USA 100 
 (6):3339.</li>
            <li class="kadov-p" value="18">Zhou 
 Q and Wong 
 WH (2004). CisModule: 
 De novo discovery 
 of cis-regulatory 
 modules by hierarchical mixture modeling. Proc 
 Natl Acad 
 Sci USA 101:12114-12119.</li>
        </ol>
        <script type="text/javascript" src="SkinSupport/MadCapBodyEnd.js">
        </script>
    </body>
</html>