<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns:MadCap="http://www.madcapsoftware.com/Schemas/MadCap.xsd" MadCap:ignoredWords="as to;INIA;website;MIAMExpress;arrays;microarrays;Affymetrix;CodeLink;dChip;RMA;PDNN;VSN;GCRMA;Quantiles;Qspline;quantile;loessScale;loessQuantile;cf;limma;vsn;datasets" MadCap:lastBlockDepth="6" MadCap:lastHeight="4133" MadCap:lastWidth="1196" MadCap:disableMasterStylesheet="true" MadCap:tocPath="Analyzing Gene Lists" MadCap:InPreviewMode="false" MadCap:RuntimeFileType="Topic" MadCap:TargetType="WebHelp" MadCap:PathToHelpSystem="../" MadCap:HelpSystemFileName="PhenoGen_Overview.xml" MadCap:SearchType="Stem">
    <head>
        <link href="SkinSupport/MadCap.css" rel="stylesheet" type="text/css" /><title>Viewing Exon-level Information</title>
        <link href="Resources/Stylesheets/PhenoGen.css" rel="stylesheet" type="text/css" />
        <script src="SkinSupport/MadCapAll.js" type="text/javascript">
        </script>
    </head>
    <body>
        <p class="MCWebHelpFramesetLink" style="display: none;"><a href="../PhenoGen_Overview_CSH.htm#Viewing_Exon_Information.htm" style="">Open topic with navigation</a>
        </p>
        <h1><a name="kanchor127"></a>Viewing Exon-level Correlations</h1>
        <p style="background-color: Transparent;">You can create an exon correlation heatmap for a specific gene, species, and tissue. The data is retrieved from multiple databases, so generating the initial graphics may take a few moments.You can display the heatmaps for two transcripts of a gene side-by-side to determine the transcript that best fits the expression correlation patterns.</p>
        <p class="References">Gene List Analysis Tools</p>
        <ol>
            <li class="Numbered" value="1">Choose <span style="font-weight: bold;">Genes List Analysis Tools</span> in the main menu.</li>
            <li class="Numbered" value="2">Click <span style="font-weight: bold;">Analyze a gene list</span>. A page displays the gene lists to 
 which you have access.</li>
            <li style="class: ;" class="Numbered" value="3">Click the gene list for 
 which you want to view gene expression data.</li>
            <li style="class: ;" class="Numbered" value="4">Click the <span style="font-weight: bold;">Exon Correlation</span>tab. The <span style="font-style: italic;">Exon-Exon Correlations
 </span>page displays.</li>
        </ol>
        <blockquote>
            <div class="MCDropDown">
                <div class="MCDropDownHead"><a class="MCDropDownHotSpot" href="javascript:void(0);" id="MCDropDownHotSpot_1459790054_0" onclick="FMCDropDown( this ); return false;"><img style="border: none;margin-right: 5px;" src="SkinSupport/DropDownClosed.gif" MadCap:altsrc="SkinSupport/DropDownOpen.gif" onload="if ( typeof( FMCPreloadImage ) == 'function' ) { FMCPreloadImage( 'SkinSupport/DropDownOpen.gif' ); }" class="MCDropDownIcon" />Take a look</a>
                </div>
                <div class="MCDropDownBody" id="MCDropDownBody_1459790054_0" style="display: none;">
                    <p>
                        <img src="Images/exonCorrelationTab.png" style="border-left-style: double;border-left-width: 4px;border-left-color: #c0c0c0;border-right-style: double;border-right-width: 4px;border-right-color: #c0c0c0;border-top-style: double;border-top-width: 4px;border-top-color: #c0c0c0;border-bottom-style: double;border-bottom-width: 4px;border-bottom-color: #c0c0c0;" />
                    </p>
                </div>
            </div>
        </blockquote>
        <p style="font-style: italic;">On the Exon-Exon Correlations page:</p>
        <ol style="list-style: decimal;list-style-type: decimal;">
            <li class="Numbered" value="1">Select the <span style="font-weight: bold;">Gene</span> 
for which you want to create an exon correlation. </li>
            <li class="Numbered" value="2">Select the <b>Species</b> and the <b>Tissue</b> to which this correlation pertains.</li>
            <li class="Numbered" value="3">Click <span style="font-weight: bold;">Get Exon Correlations</span>. The exon correlations display.</li>
        </ol>
        <blockquote>
            <div class="MCDropDown">
                <div class="MCDropDownHead"><a class="MCDropDownHotSpot" href="javascript:void(0);" id="MCDropDownHotSpot_1459790054_1" onclick="FMCDropDown( this ); return false;"><img style="border: none;margin-right: 5px;" src="SkinSupport/DropDownClosed.gif" MadCap:altsrc="SkinSupport/DropDownOpen.gif" onload="if ( typeof( FMCPreloadImage ) == 'function' ) { FMCPreloadImage( 'SkinSupport/DropDownOpen.gif' ); }" class="MCDropDownIcon" />Take a look</a>
                </div>
                <div class="MCDropDownBody" id="MCDropDownBody_1459790054_1" style="display: none;">
                    <p>
                        <img src="Images/viewExonCorrelation_Full.png" style="border-left-style: double;border-left-width: 4px;border-left-color: #c0c0c0;border-right-style: double;border-right-width: 4px;border-right-color: #c0c0c0;border-top-style: double;border-top-width: 4px;border-top-color: #c0c0c0;border-bottom-style: double;border-bottom-width: 4px;border-bottom-color: #c0c0c0;" />
                    </p>
                </div>
            </div>
        </blockquote>
        <ol MadCap:continue="true">
            <li value="4"><span style="font-weight: bold;font-family: 'Comic Sans MS', cursive;color: #000080;">OPTIONAL:</span>Display two heat maps by selecting both a Left and Right Transcript in the In the <i>Gene/Transcript</i> <![CDATA[ ]]>section at the top left. </li>
            <li value="5"><span style="font-weight: bold;font-family: 'Comic Sans MS', cursive;color: #000080;">OPTIONAL:</span> Hover over the schematic on top of the heatmap or click on individual exons in the schematic to find out more information about individual exons and the probe sets that align to them.</li>
        </ol>
        <p>By default, the first graphic, which is generated through the University of California Santa Cruz Genome Browser (<a href="http://genome.ucsc.edu/">http://genome.ucsc.edu/</a>), displays the probe sets from the exon array and the annotated Ensembl transcripts (i.e., isoforms) for the input gene.   </p>
        <p>
            <img src="Images/viewExonCorrelation_Graph.png" style="border-left-style: double;border-left-width: 4px;border-left-color: #c0c0c0;border-right-style: double;border-right-width: 4px;border-right-color: #c0c0c0;border-top-style: double;border-top-width: 4px;border-top-color: #c0c0c0;border-bottom-style: double;border-bottom-width: 4px;border-bottom-color: #c0c0c0;" /> <![CDATA[ ]]></p>
        <p>The exons are color-coded based on their level of annotation according to Affymetrix.  The red probe sets are included in the core set of probe sets and indicate high confidence in the annotation.  The blue probe sets are included in the extended set of probe sets according to Affymetrix, and the green probe sets are from the full set of probe sets and indicate the least amount of confidence in the annotation.  The next track assigns a number to each annotated exon of the gene to aid in interpretation of the heatmap that follows.
</p>
        <p>By default, the second graphic is a heatmap that displays the correlations among exons of a given transcript of the gene.</p>
        <p>
            <img src="Images/viewExonCorrelation_Heatmap.png" style="border-left-style: double;border-left-width: 4px;border-left-color: #c0c0c0;border-right-style: double;border-right-width: 4px;border-right-color: #c0c0c0;border-top-style: double;border-top-width: 4px;border-top-color: #c0c0c0;border-bottom-style: double;border-bottom-width: 4px;border-bottom-color: #c0c0c0;" />
        </p>
        <p> Each row and column represents a probe set that passed the filtering criteria specified at the top right side of the page. Along the top of the heatmap is a schematic of the exon structure of the transcript.  The boxes represent exons of the transcript and are numbered, as in the top figure.  Grey boxes are exons that are probed by the array, met the filtering criteria, and are displayed in the heatmap.  Exons that do not satisfy these criteria are colored based on the reason why they are not included in the heatmap.

</p>
        <ul>
            <li value="1"><b>No Probesets (Red)</b> indicates that the Affymetrix array does not contain any probe sets that target this exon.
</li>
            <li value="2"><b>Masked (Light Blue)</b> indicates that the probe set was removed due to poor intergrity of its probes (i.e., the probes aligned to many regions of the genome or targeted a region of the genome that harbored a known SNP)
.</li>
            <li value="3"><b>Heritability (Blue)</b> indicates that the probe set had a low (according to filtering criteria) heritability in the RI panel. 
</li>
            <li value="4"><b>Annotation Confidence (Orange)</b> indicates that probe sets that align to this exon are not included due the annotation confidence criteria selected for filtering.
	</li>
            <li value="5"><b>Not DABG (Green)</b> indicates that the probe sets that align to this exon are not expressed above background according to the DABG criteria selected for filtering
.</li>
            <li value="6"><b>Multiple Reasons (Yellow)</b> indicates that the exon was filtered out for more that one reason, e.g., low heritability and not DABG

</li>
        </ul>
        <p>In the top right section, choose the probe sets to display for the heatmap based on the:          <![CDATA[ ]]></p>
        <ul>
            <li value="1"> Annotation Confidence from Affymetrix (Core, Extended, or Full).</li>
            <li value="2">Heritability of the probe set’s expression in the RI panel.</li>
            <li value="3">Proportion of samples with expression values above background (Detected Above Background–DABG). </li>
        </ul>
        <p>Choose the <b>Adv. Display Options</b> tab to show advanced display options that allow you to include probe sets that align to an intronic region of the gene or align to the opposite strand from which the gene is coded. </p>
        <p>
            <img src="Images/viewExonCorrelation_Settings.png" style="border-left-style: double;border-left-width: 4px;border-left-color: #c0c0c0;border-right-style: double;border-right-width: 4px;border-right-color: #c0c0c0;border-top-style: double;border-top-width: 4px;border-top-color: #c0c0c0;border-bottom-style: double;border-bottom-width: 4px;border-bottom-color: #c0c0c0;" />
        </p>
        <script type="text/javascript" src="SkinSupport/MadCapBodyEnd.js">
        </script>
    </body>
</html>