<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns:MadCap="http://www.madcapsoftware.com/Schemas/MadCap.xsd" MadCap:lastBlockDepth="5" MadCap:lastHeight="2646" MadCap:lastWidth="1207" MadCap:disableMasterStylesheet="true" MadCap:tocPath="Analyzing Microarrays" MadCap:InPreviewMode="false" MadCap:RuntimeFileType="Topic" MadCap:TargetType="WebHelp" MadCap:PathToHelpSystem="../" MadCap:HelpSystemFileName="PhenoGen_Overview.xml" MadCap:SearchType="Stem">
    <head>
        <link href="SkinSupport/MadCap.css" rel="stylesheet" type="text/css" /><title>Viewing Datasets</title>
        <link href="Resources/Stylesheets/PhenoGen.css" rel="stylesheet" type="text/css" />
        <script src="SkinSupport/MadCapAll.js" type="text/javascript">
        </script>
    </head>
    <body>
        <p class="MCWebHelpFramesetLink" style="display: none;"><a href="../PhenoGen_Overview_CSH.htm#Viewing_Datasets.htm" style="">Open topic with navigation</a>
        </p>
        <h1 class="pgBrkBefore"><a name="kanchor16"></a>Viewing Datasets</h1>
        <p>You can see the collection of all your datasets at any time. The page 
 that displays your datasets shows you the stage that each dataset is in (e.g., 
 quality control completed, grouped and normalized, 
 etc.) </p>
        <ol>
            <li class="Numbered" value="1">Click the <span style="font-weight: bold;">Analyze 
 Microarrays</span> tab. The <span style="font-style: italic;">Analyze 
 Microarrays </span>page displays.</li>
            <li class="Numbered" value="2">Click <span style="font-weight: bold;">Analyze 
 microarray data</span> in the <span style="font-style: italic;">What would 
 you like to do</span> section. A page displays your datasets.</li>
        </ol>
        <p>At the top of the page, you can click the <span style="font-weight: bold;">Create 
 Dataset</span> option if you want to retrieve and select arrays and finalize 
 them into a new dataset, or click the <b>Upload Arrays</b> option to upload your own arrays and create a dataset. The page provides four grouped and normalized 
 "Public" 
 datasets for you to analyze and save new gene lists. 
 &#160;<a href="#publicDatasets" target="" title="" alt="" class="MCXref" xrefformat="See {quote}{paratext}{quote}">See "Public Datasets"</a> for information 
 about public datasets.</p>
        <p>Another table displays after the Public Datasets table and shows your 
 "Private" datasets. After you finalize a dataset, it becomes 
 part of the <span style="font-style: italic;">My Private Datasets</span> 
 table, where your progress on that dataset is denoted in the columns:</p>
        <ul>
            <li class="Bullet" value="1"><span style="font-weight: bold;">QC 
 Complete</span>: The word "Run" displays in the QC Complete 
 column until you perform quality control checks. At that time, "Review 
 Results" displays in the QC Complete column. You must review your 
 quality control results and approve them before a checkmark displays in 
 this column. </li>
            <li class="Bullet" value="2"><span style="font-weight: bold;">Arrays 
 Grouped and Normalized</span>: The word "Run" displays in the 
 Arrays Grouped and Normalized column until you create a group and normalize 
 the data based on that grouping. You can create multiple array groupings 
 and normalize each group many different ways. Each normalized grouping 
 is saved as a new dataset "version". After you create and normalize 
 your grouped arrays, a checkmark displays in this column.</li>
            <li class="Bullet" value="3"><b>Phenotype Data</b>:&#160;A magnifying glass displays in this column if there is phenotype data for the dataset.</li>
        </ul>
        <h5 style="margin-top: 12pt;">Results section</h5>
        <ul>
            <li value="1"><b>Quality Control Results</b>: A magnifying glass displays in this column if there are quality control results for the dataset.</li>
            <li class="Bullet" value="2"><b>Cluster Results</b>: A magnifying glass displays in this column if there are cluster results for the dataset.</li>
            <li class="Bullet" value="3"><span style="font-weight: bold;">Gene 
 Lists Saved</span>: A checkmark displays in this column after you analyze 
 the normalized dataset and save the resultant gene list.</li>
        </ul>
        <p class="Numbered" style="margin-left: 0px;">You can:</p>
        <ul>
            <li class="Bullet" value="1">Click a dataset to view that 
 dataset in its current state of processing.</li>
            <li class="Bullet" value="2">Click the <span style="font-weight: bold;">View</span> 
 link in the <span style="font-weight: bold;">Details</span> column to 
 view dataset 
 details 
 such as name, description, organism, arrays in dataset, and more. <a href="Viewing_Dataset_Details.htm" target="" title="" alt="" class="MCXref" xrefformat="See {quote}{paratext}{quote}">See "Viewing Dataset Details"</a> for more 
 information.</li>
            <li class="Bullet" value="3">Click the <span style="font-weight: bold;">Delete 
 </span>icon <img src="Images/deleteIcon.gif" style="border: none;width: 27px;height: 20px;float: none;border-style: none;border-style: none;" width="27" height="20" border="0" />to delete a dataset. </li>
            <li class="Bullet" value="4">Click the <span style="font-weight: bold;">Download</span> 
 icon <img src="Images/downloadIcon.gif" style="border: none;width: 20px;height: 23px;float: none;border-style: none;border-style: none;" width="20" height="23" border="0" /> to download a dataset. </li>
        </ul>
        <blockquote>
            <div class="MCDropDown">
                <div class="MCDropDownHead"><a class="MCDropDownHotSpot" href="javascript:void(0);" id="MCDropDownHotSpot_2959807513_0" onclick="FMCDropDown( this ); return false;"><img style="border: none;margin-right: 5px;" src="SkinSupport/DropDownClosed.gif" MadCap:altsrc="SkinSupport/DropDownOpen.gif" onload="if ( typeof( FMCPreloadImage ) == 'function' ) { FMCPreloadImage( 'SkinSupport/DropDownOpen.gif' ); }" class="MCDropDownIcon" />Take a look</a>
                </div>
                <div class="MCDropDownBody" id="MCDropDownBody_2959807513_0" style="display: none;">
                    <p>
                        <img src="Images/viewingDatasets.png" style="border-left-style: Double;border-left-width: 4px;border-right-style: Double;border-right-width: 4px;border-top-style: Double;border-top-width: 4px;border-bottom-style: Double;border-bottom-width: 4px;border-left-color: #c0c0c0;border-right-color: #c0c0c0;border-top-color: #c0c0c0;border-bottom-color: #c0c0c0;float: none;" border="4" />
                    </p>
                </div>
            </div>
        </blockquote>
        <h2><a name="publicDatasets"></a>Public Datasets</h2>
        <p>The Public datasets available for analysis on the PhenoGen website are 
 pre-compiled groupings of gene expression data for various strains 
 of inbred and recombinant inbred mice and rats. These datasets are available 
 for all types of analysis by any registered users but may be most useful 
 for performing correlation analysis with phenotype data. These datasets 
 are normalized using the most common normalization techniques and have 
 already had quality control checks run. Additionally, datasets created using the Affymetrix Exon arrays have been adjusted for batch effects using an empirical Bayes method (Johnson et al 2007). The normalized data or raw data 
 can be downloaded from the <a href="DownloadResources.htm">Download Resources</a> page.</p>
        <h3>Inbred Mice</h3>
        <p>The whole brain gene expression dataset for the inbred mice includes 
 20 inbred strains. Each strain has four to seven biological replicates 
 for a total of 90 individual arrays. The whole brain mRNA for each naive 
 10-12 week old male mouse was hybridized to a separate array, i.e., no 
 pooling of samples. </p>
        <p>The inbred mouse data was normalized nine different ways. Five of the 
 normalization methods are available on the website. For the other four 
 versions, a probe mask was created to eliminate probes whose sequences 
 did not match to the NCBI m37 Build, matched the genome in multiple places, 
 or harbored a SNP between any of the 19 strains where genotype data is 
 available at the Imputed Genotype Resource from the Jackson Laboratory; 
 <a href="http://cgd.jax.org/datasets/popgen/imputed.shtml" target="_blank">http://cgd.jax.org/datasets/popgen/imputed.shtml</a> 
 (129P3/J is not available). &#160;Entire 
 probe sets were eliminated if less than four associated probes remained. 
 The version using the probe mask and the RMA normalization method is the 
 RECOMMENDED version. </p>
        <h3 class="pgBrkBefore">BXD Recombinant Inbred Mice</h3>
        <p>The whole brain gene expression dataset for the BXD recombinant inbred 
 mice includes 30 recombinant inbred strains and the two parental strains 
 (C57BL/6J and DBA/2J). 
 Each strain has four to seven biological replicates for a total of 172 
 individual arrays. The whole brain mRNA for each naive 10-12 week old 
 male mouse was hybridized to a separate array, i.e., no pooling of samples. 
 </p>
        <p>The BXD data was normalized nine different ways. Five of the normalization 
 methods use all of the probes in the dataset. A probe mask was created 
 for the other four versions to eliminate probes whose sequences did not 
 match to the NCBI m37 Build, matched the genome in multiple places, or 
 harbored a SNP between any of the 19 inbred mouse strains included in 
 the public dataset (according to Imputed Genotype Resource from the Jackson 
 Laboratory; <a href="http://cgd.jax.org/datasets/popgen/imputed.shtml" target="_blank">http://cgd.jax.org/datasets/popgen/imputed.shtml</a>). 
 Entire probe sets were eliminated if less than four associated probes 
 were eliminated. The version using the probe mask and the RMA normalization 
 method is the RECOMMENDED version. </p>
        <p>For the eQTL analysis of this data set, a slightly different mask was used.  Instead of eliminated probes with SNPs between the 19 inbred strains, probes were eliminated if they contained a known SNP between the two BXD parental strains, based on whole genome sequence data from the Sanger Institute (Keane et al 2011). Expression values were normalized and summarized into probesets using RMA.  MAS5 was used to evaluate if expression level measurements were above background noise (present, absent, or marginal).  If a probeset did not have at least one present call throughout all samples, the probeset was dropped from the data set.  Of the 41,581 probesets retained after masking, 30,031 probesets remained after filtering by present/absent calls.  Data were thoroughly examined for batch effects related to processing.  The microarrays were run over a year and a half period, resulting in 15 batches.  Both batches and strains contribute to non-random data distribution and a new method for removing batch effects, while retaining strain effects, was used (personal communication, Evan Johnson, Boston University) on the set of 30,031 probesets detected above background.  This method combines a simple rank test and a Bayesian hierarchical framework similar to the empirical Bayes method, <i>Combating Batch Effects When Combining Batches of Gene Expression Microarray Data</i> (ComBat) (Johnson et al., 2007).  This version of the data is available in the <a href="DownloadResources.htm">Download Resources</a> section.
 </p>
        <h3>BXD Recombinant Inbred and Inbred Mice</h3>
        <p>This expression data set represents a combination of the two datasets 
 previously mentioned. In this dataset there are a total of 50 strains 
 (C57BL/6J and DBA/2J 
 are in both of the previous sets) and 253 individual arrays. See the preceding 
 topic for details on "masked" versions.</p>
        <h3>LXS Recombinant Inbred Mice</h3>
        <p>The whole brain gene expression dataset for the LXS recombinant inbred male mice includes 59 recombinant inbred strains (one strain (LXS49) was eliminated due to unresolved questions about true strain origin) and two parental strains (ILS and ISS). Each strain has three to six biological replicates, for a total of 342 individual arrays that passed quality control standards. In addition, to control for batch effects, C57BL/6J mice were hybridized to arrays and included in every batch (35 individual arrays), and DBA/2J mice were included in a few of the final batches (9 arrays). The whole brain mRNA for each naive 10-12 week old male mouse was hybridized to a separate Affymetrix Mouse Exon Array 1.0 ST, i.e., no pooling of samples.</p>
        <p> Individual probes were eliminated prior to normalization if their sequence did not match any part of the NCBI m37 Build of the mouse genome, if their sequence matched multiple locations in the mouse genome, or if the location in the genome that the probe did match contain a SNP between any of the 19 strains in the public Inbred Mice dataset where genotype data is available at the Imputed Genotype Resource from the Jackson Laboratory; <a href="http://cgd.jax.org/datasets/popgen/imputed.shtml" target="_blank">http://cgd.jax.org/datasets/popgen/imputed.shtml</a> (same mask that is implemented on PhenoGen). Entire probe sets were eliminated if less than three of the original probes remained after filtering. Arrays were examined for quality, and arrays that did not meet quality standards were eliminated. </p>
        <p> Data from individual probes was normalized using RMA and summarized either into the full set of transcript cluster or the core set of transcript clusters.  In addition, RMA values for the full set of individual probe sets is available for download from the resource page, but is not available for analysis on PhenoGen at this time. </p>
        <p>Each data set was adjusted for batch effects using the empirical Bayes method outlined by Johnson et al (2007).  After batch effects adjustment, C57BL/6J and DBA/2J arrays were dropped from the data set.  The version using the probe mask and the RMA normalization method on the core transcript clusters is the <i style="font-weight: bold;">recommended</i> version and was used for calculation of eQTLs.</p>
        <h3>HXB/BXH Recombinant Inbred Rats</h3>
        <p>The whole brain gene expression dataset for the HxB/BxH 
 recombinant inbred rats on the CodeLink Whole Genome Rat Array includes data from 26 recombinant inbred strains, 
 the two parental strains (SHR/Ola 
 and BN-Lx/Cub), 
 and the SHR-Lx/Cub 
 strain. The whole brain mRNA of four to seven naive 12-14 week old male 
 rats from each strain were hybridized to separate CodeLink Whole Genome 
 rat arrays (one rat per array) for a total of 139 arrays. </p>
        <p>In addition to the five normalization versions available on the website, 
 an "eQTL version" of the dataset that was used for all HXB/BXH 
 rat eQTL calculations is available. This version was obtained by first 
 removing probes from the datasets if they were one of the negative or 
 positive controls placed on the array by the manufacturer. Next, individual 
 values were eliminated based on the quality flags assigned by the CodeLink 
 Expression Analysis Software. Values were eliminated if they were flagged 
 as M (spot was identified to be defective through image inspection at 
 manufacturing), C (spot has a high level of background contamination), 
 I (spot has an irregular shape), or S (spot has a high number of saturated 
 pixels). Values were retained if they were flagged G (spot is good) or 
 L (spot is below local background noise). Also, to be able to take the 
 log base 2 transformation of the background-adjusted intensity values, 
 all background-adjusted intensity values below zero were replaced with 
 the value 0.00001. The data was then normalized using a cyclic LOESS procedure 
 executed in R to account for the missing intensity values. </p>
        <p>The HXB/BXH recombinant inbred panel also has four data sets available on transcription levels from the Affymetrix Rat Exon array.  Data was collected on whole brain, left ventricle (heart), liver, and brown adipose tissue (BAT) of 21 HXB/BXH RI strains (only 19 RI strains included in the BAT tissue data set) and 6 related inbred strains.  Each strain has three to four biological replicates for a total of 108 individual arrays from brain, 105 arrays from heart, 106 arrays from liver, and 96 arrays from brown adipose tissue that passed quality control standards. The mRNA for each naive 10 week old male rat was hybridized to a separate Affymetrix Rat Exon Array 1.0 ST, i.e., no pooling of samples.</p>
        <p>Individual probes were eliminated prior to normalization if their sequence did not match any part of the RGSC version 3.2 of the rat genome, if their sequence matched multiple locations in the mouse genome, or if the location in the genome that the probe did match contain a SNP between the Brown Norway (BN/SsNHsdMcwi) inbred strains (reference strain) and the spontaneously hypertensive rat (SHR/OlaIpcv) strain that was recently sequenced (Atanur et al 2010) using next generation sequencing or a SNP detected in DNA sequencing of the BN-Lx/CubPrin and SHR/OlaIpcvPrin strains (same mask that is implemented on PhenoGen). DNA sequence data for the BN/SsNHsdMcwi and SHR/OlaIpcv was downloaded directly from the Ensembl ftp site at:&#160;<a href="ftp://ftp.ebi.ac.uk/pub/databases/ensembl/snp/rat/shr/" target="_blank">ftp://ftp.ebi.ac.uk/pub/databases/ensembl/snp/rat/shr/</a>.   </p>
        <p>For the 4,022,111 original probes, 604,601 were removed (472,072 did not map uniquely to the genome; 132,529 probes contained a SNP).  Entire probe sets were eliminated if less than three of the original probes remained after filtering.  Arrays were examined for quality and arrays that did not meet quality standards were eliminated. </p>
        <p> Data from individual probes was normalized using RMA and summarized either into the full set of transcript cluster or the core set of transcript clusters.  In addition, RMA values for the full set of individual probe sets is available for download from the resource page, but is not available for analysis on PhenoGen at this time.</p>
        <p>Each data set was adjusted for batch effects using the empirical Bayes method outlined by Johnson et al. (2007). The version using the probe mask and the RMA normalization method on the core transcript clusters is the <i style="font-weight: bold;">recommended</i> version and was used for calculation of eQTLs.</p>
        <p class="References"><b>References</b>
        </p>
        <ol>
            <li value="1">Johnson WE, Li C, and Rabinovic A (2007).  Adjusting batch effects in microarray expression data using empirical Bayes methods. Biostatistics 8(1): 118-127.<br /><br /></li>
            <li value="2">Atanur SS, Birol I, Guryev V, Hirst M, Hummel O, Morrissey C, Behmoaras J, Fernandez-Suarez XM, Johnson MD, McLaren WM, Patone G, Petretto E, Plessy C, Rockland KS, Rockland C, Saar K, Zhao Y, Carninci P, Flicek P, Kurtz T, Cuppen E, Pravenec M, Hubner N, Jones SJ, Birney E, Aitman TJ (2010). The genome sequence of the spontaneously hypertensive rat: Analysis and functional significance. Genome Research 20(6):791-803.</li>
        </ol>
        <script type="text/javascript" src="SkinSupport/MadCapBodyEnd.js">
        </script>
    </body>
</html>