<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns:MadCap="http://www.madcapsoftware.com/Schemas/MadCap.xsd" MadCap:lastBlockDepth="4" MadCap:lastHeight="1199" MadCap:lastWidth="1203" MadCap:disableMasterStylesheet="true" MadCap:tocPath="Analyzing Microarrays|Creating Datasets|Quality Control Checks|Assessing Affymetrix Data Quality" MadCap:InPreviewMode="false" MadCap:RuntimeFileType="Topic" MadCap:TargetType="WebHelp" MadCap:PathToHelpSystem="../" MadCap:HelpSystemFileName="PhenoGen_Overview.xml" MadCap:SearchType="Stem">
    <head>
        <link href="SkinSupport/MadCap.css" rel="stylesheet" type="text/css" /><title>Within-Array Checks for Affymetrix Exon Arrays</title>
        <link href="Resources/Stylesheets/PhenoGen.css" rel="stylesheet" type="text/css" />
        <script src="SkinSupport/MadCapAll.js" type="text/javascript">
        </script>
    </head>
    <body>
        <p class="MCWebHelpFramesetLink" style="display: none;"><a href="../PhenoGen_Overview_CSH.htm#Within-Array_Checks_for_Affymetrix_Exon_Arrays.htm" style="">Open topic with navigation</a>
        </p>
        <h1><a name="kanchor30"></a>Within-Array Checks for Affymetrix Exon Arrays</h1>
        <p style="background-color: Transparent;">The within-array quality control 
 checks are examined using data from quality control reports produced by 
 the Affymetrix Power Tools and graphics from the Bioconductor package 
 <span style="font-style: italic;">Simpleaffy</span>. Quality control measures 
 are based on an RMA normalization of the core transcript clusters. There 
 are four checks that are examined:</p>
        <ul>
            <li class="Bullet" value="1"><span style="font-weight: bold;"><a name="Average_Background"></a>Average Background</span> &#160;The 
 average raw intensity value across probes used to calculated background. 
 The average is calculated on intensity values prior to any normalization. 
 There is no relevant threshold for this value. Look for consistency across 
 arrays.</li>
            <li class="Bullet" value="2"><a name="Percent_Present"></a><span style="font-weight: bold;">Percent Present</span> &#160;The 
 proportion of all probe sets (transcript clusters) with an intensity value 
 above detection limits (p-value &lt; 0.01). Although the percent of transcript 
 clusters above detection limits is highly dependent on each specific experiment 
 with respect to the number of genes you expected to be expressed, an extremely 
 low value raises suspicion about the quality of an array. It is expected 
 that duplicate arrays have similar percent-missing levels.</li>
            <li class="Bullet" value="3"><a name="Scaling_Factors"></a><span style="font-weight: bold;">Pseudo Scaling Factors</span> &#160;Represents 
 the replicated scaling factor from the quality control measure for the 
 3’ expression arrays. It is the ratio of the average raw intensity value 
 of all probes on the array compared to the average raw intensity value 
 across all arrays in the experiment. This value gives a general idea of 
 how ‘dim’ or ‘bright’ an array is. &#160;Most 
 minor discrepancies among arrays are eliminated with proper normalization.</li>
            <li class="Bullet" value="4"><span style="font-weight: bold;"><a name="AUC"></a>AUC for Distinguishing Positive and Negative Controls</span> 
 &#160;The positive 
 and negative controls are used in a receiver operating characteristic 
 (ROC) analysis to assess the array's ability to distinguish between the 
 two, based on signal intensity. The area under the curve (AUC) is a descriptive 
 measure to assess this accuracy. An AUC of 1 indicates the perfect separation 
 of positive and negative controls based on signal intensity, while an 
 AUC of 0.5 indicates that signal intensity cannot be used to distinguish 
 between the two types of probes. The AUC can also be calculated based 
 on detection above background measures.</li>
        </ul>
        <p style="background-color: Transparent;">These measures are displayed 
 in a graphic similar to the one generated by the <span style="font-style: italic;">Simpleaffy</span> 
 package from Bioconducter for the 3’ Affymetrix arrays. </p>
        <div class="MCDropDown">
            <div class="MCDropDownHead"><a class="MCDropDownHotSpot" href="javascript:void(0);" id="MCDropDownHotSpot_3797520184_0" onclick="FMCDropDown( this ); return false;"><img style="border: none;margin-right: 5px;" src="SkinSupport/DropDownClosed.gif" MadCap:altsrc="SkinSupport/DropDownOpen.gif" onload="if ( typeof( FMCPreloadImage ) == 'function' ) { FMCPreloadImage( 'SkinSupport/DropDownOpen.gif' ); }" class="MCDropDownIcon" />Take a look</a>
            </div>
            <div class="MCDropDownBody" id="MCDropDownBody_3797520184_0" style="display: none;">
                <p>
                    <img src="Images/exon_summary.png" style="border-left-style: Double;border-left-width: 4px;border-right-style: Double;border-right-width: 4px;border-top-style: Double;border-top-width: 4px;border-bottom-style: Double;border-bottom-width: 4px;border-left-color: #c0c0c0;border-right-color: #c0c0c0;border-top-color: #c0c0c0;border-bottom-color: #c0c0c0;float: none;width: 600px;height: 600px;" width="600" height="600" border="4" />
                </p>
            </div>
        </div>
        <p style="background-color: Transparent;"><span style="background-color: Transparent;">Along 
 the left side of the graph are the names of the CEL files that are included 
 in the analysis. The next column has two numbers per CEL file: the top 
 number is the <span style="font-style: italic;">percent present</span> 
 and the bottom number is the <span style="font-style: italic;">average 
 background</span> for that CEL file.</span>
        </p>
        <p style="background-color: Transparent;">The percent present values are 
 heavily dependent on the type of sample used on the array. If the same 
 type of tissue is used in all samples, the percent present values should 
 be similar across arrays. However, if you have multiple tissue types such 
 as liver and brain, the percent present values can vary substantially 
 between these tissues. By default, the percent present values are displayed 
 in red if there is a spread greater than 10% between the lowest and highest 
 values within the experiment. <span style="background-color: Transparent;">It 
 is expected that the average background values across arrays should be 
 similar. If they vary by more than 20 units, all values are given in red.</span></p>
        <p style="background-color: Transparent;">The solid dots that are attached 
 to a horizontal line originating from the solid vertical line through 
 1 represent the pseudo-scaling factors (indicators of how much RNA was 
 hybridized to the array) for each array. &#160;The 
 blue shaded region is the area 30% above and 30% below the experiment 
 average. Values that fall outside this region are displayed in red.</p>
        <p style="background-color: Transparent;">The open circles and open triangles 
 represent the AUC for distinguishing positive and negative controls respectively, 
 based on intensity values and detection above background values. According 
 to Affymetrix, “values between 0.80 and 0.90 are typical” (<span style="font-style: italic;">Quality 
 Assessment of Exon and Gene Arrays, 2007</span>) for AUCs based on signal 
 intensity. AUC values based on detection above background tend to have 
 values between 0.75 and 0.85.</p>
        <script type="text/javascript" src="SkinSupport/MadCapBodyEnd.js">
        </script>
    </body>
</html>